File: cpp/bc2026/src/bestsofar/Utils/FastSet.cpp
#include "Utils/FastSet.hpp"

namespace Utils {

bool FastSet::add(char16_t value) {
    std::u16string str(1, value);
    if (values.find(str) == std::u16string::npos) {
        values.append(str);
        return true;
    }

    return false;
}

bool FastSet::contains(char16_t value) {
    return values.find(std::u16string(1, value)) != std::u16string::npos;
}

bool FastSet::add(MapLocation location) {
    return add(encodeLocation(location));
}

bool FastSet::contains(MapLocation location) {
    return contains(encodeLocation(location));
}

char16_t FastSet::encodeLocation(MapLocation location) {
    return (char16_t)((location.x << 6) | location.y);
}

}  // namespace Utils

File: py/bc2026/src/bestsofar/Utils/FastSet.py
# Port of bestsofar/Utils/FastSet.java. See py/CONVENTIONS.md.
from api import *


class FastSet:
    def __init__(self):
        self.values = ""  # Java: private StringBuilder values

    # Java: two overloads, add(char) and add(MapLocation) - the MapLocation
    # one encodes the location and delegates to the char one.
    def add(self, value):
        if isinstance(value, MapLocation):
            return self._addChar(self.encodeLocation(value))
        return self._addChar(value)

    def _addChar(self, value):
        s = chr(value)  # Java: String str = String.valueOf(value)
        if self.values.find(s) == -1:
            self.values += s
            return True

        return False

    # Java: contains(char) / contains(MapLocation) overloads, same dispatch.
    def contains(self, value):
        if isinstance(value, MapLocation):
            value = self.encodeLocation(value)
        return self.values.find(chr(value)) > -1

    def encodeLocation(self, location):
        # Java: (char) cast truncates to 16 bits
        return ((location.x << 6) | location.y) & 0xFFFF
